### ISSUE TITLE
Heads up notification not working

### ISSUE THREAD

**Issue by balajikarthiq** [fechanique/cordova-plugin-fcm#365]
Heads up notification is not working, even setting the priority as "High" and checked the file MyFirebaseMessagingService.java.

The sendNotification method is commented in onMessageReceived method.

**Comment by CowboyCode** [comment]
It is working in foreground but it doesn't make a sound. The message is received, like you would click on it, receiving it in background. 

In your code, onNotification, you can display the message in a modal for example.
I save all notifications in sqlite and have a screen that shows all messages. In the onNotification, I redirect to the notification screen.

FYI

**Comment by munyanyika24** [comment]
I am getting notification on Android but on iOS the onNotification method is not firing at all.

I am running iOS 10.3, in console, I see messages that everything is initialized and token is registered but when a send a notification via the firebase console, I receive it on Android but not on iOS. The plugin version i am running is 2.1.2 (latest one)

Please assist

**Comment by CowboyCode** [comment]
Did you set up your ios certificates in firebase?

**Comment by 1N50MN14** [comment]
@CowboyCode I'm using your fork and all works for me on iOS excepting when sending data notifications, the xcode log shows:

`FIRMessaging received data-message, but FIRMessagingDelegate's-messaging:didReceiveMessage: not implemented`

There's an open PR on this issue but it unfortunately conflicts with your modifications I was just wondering if you somehow got this working for you.. thank you (and for spending time trying to assist others on this repo)

**Comment by CowboyCode** [comment]
I will test it on Monday morning and let you know

**Comment by 1N50MN14** [comment]
@CowboyCode Thank you

**Comment by blumanski** [comment]
@1N50MN14 It works for me, see my Xcode log below.
I have it running on an iphone 6 and have my app targeted to 9.0
May remove the plugin and add it back helps?



    2017-08-28 11:01:17.950265+1000 Projectname[347:20457] Message ID 1: 
    0:1503882076935879%7df1ea3a7df1ea3a
    2017-08-28 11:01:17.951057+1000 Projectname[347:20457] {
    aps =     {
        alert =         {
                body = dsfdsfds;
                title = sdfdsfdsf;
            };
            category = "FCM_PLUGIN_ACTIVITY";
            sound = default;
        };
        channels = "[\"Test\"]";
        created = 1503882075000;
        doc = "";
        "gcm.message_id" = "0:1503882076935879%7df1ea3a7df1ea3a";
        header = sdfdsfdsf;
        message = dsfdsfds;
        page = "";
        type = page;
    }

    2017-08-28 11:01:17.951407+1000 Projectname[347:20457] 
    stringByEvaluatingJavaScriptFromString 
    FCMPlugin.onNotificationReceived({"doc":"","header":"sdfdsfdsf","channels":"[\"Test\"]","aps":
    {"alert":{"title":"sdfdsfdsf","body":"dsfdsfds"},"sound":"default","category":"FCM_PLUGIN_ACTIVITY"},"message":"dsfdsfds","created":"1503882075000","type":"page","page":"","gcm.message_id":"0:1503882076935879%7df1ea3a7df1ea3a"});

I receive the messages as below

    if (typeof(FCMPlugin) != 'undefined') {
        FCMPlugin.onNotification(function(data){
	    console.log(data)
	    ….

Then I pop the message up using

    $cordovaToast
        .show(
            data.header + "\n" + data.message,
            'long',
            'center'
      )

After that I redirect to a notification page which shows all received notifications, I save those in sqlite.

Without the implemented code, you would not know that you got a message, there is noise when the app is in foreground.

**Comment by 1N50MN14** [comment]
@blumanski Thank you, this works for you because you're sending it as notification message and not as a data message, the above works for me as well. I'm sending a data message in accordance to FCM data specs (https://firebase.google.com/docs/cloud-messaging/concept-options) whereby you don't specify the `title`and `body` attributes, instead a `data` object which is passed silently to the app without triggering a notification alert.

**Comment by blumanski** [comment]
@1N50MN14 lol, I have to many github accounts...

Ah, ok, I just add my data to the notification, FCM does not display the data, just the header and body when the app is in the background. The data has still to be done in the code.

Can't you send it as notification and send the data with it? I add heaps of data to it

**Comment by 1N50MN14** [comment]
@blumanski :)

Yes I can send a notification with data with it but then it would be displayed to the user by the OS whereas the intention of sending a data message is to let the app deal with the processing / handling of the data.

In my case, I'm using a push message to notify users with things such as "someone liked your post" and data message to instruct the app to perform a certain task that does not involve user interaction.

To test this, remove the `title` and `body` attributes from your message; instead only include a `data` object, you'll see the error thrown in xcode

**Comment by blumanski** [comment]
I will check that out

**Comment by blumanski** [comment]
@1N50MN14 
I think it needs another upgrade of the firebase ios sdk.
I give it a shot

https://github.com/firebase/quickstart-cpp/issues/9

**Comment by 1N50MN14** [comment]
Yes this is it! Thank you for looking into it despite not needing it yourself, very much appreciated @blumanski

**Comment by CowboyCode** [comment]
@1N50MN14 
No worries :-)
I did the upgrade to 4.1.0 firebase ios sdk, but it is still not working.
You can test it here
https://github.com/CowboyCode/cordova-plugin-fcm

Haven't done a pull request yet

**Comment by 1N50MN14** [comment]
Oh bummer, will test it right away one sec

**Comment by 1N50MN14** [comment]
@CowboyCode Yup, same error:
`<Warning> [Firebase/Messaging][I-FCM002019] FIRMessaging received data-message, but FIRMessagingDelegate's-messaging:didReceiveMessage: not implemented`

hmmmm this is strange..  did you download the release from https://firebase.google.com/docs/cpp/setup?

**Comment by blumanski** [comment]
Na, the ios version, 4.1.0. 
https://firebase.google.com/download/ios

The error must come from the plugin code, I am not a object-c developer :-)

**Comment by 1N50MN14** [comment]
@blumanski ahh bummer.. neither am I, this language is just too weird.. ;)

**Comment by blumanski** [comment]
@1N50MN14 it looks like everything is in the code, I guess it would take somebody 5 minutes to add this :-) It would take me probably 10 hours lol, if it even would work :-)

**Comment by CowboyCode** [comment]
@1N50MN14 
There is actually a pull request for the same issue #349, I didn't see that.
I had only ten minutes and just pushed it up, may works right away, otherwise I will take more time to do it. If you want to try it, it's on https://github.com/CowboyCode/cordova-plugin-fcm

Cheers

**Comment by 1N50MN14** [comment]
Thank you @CowboyCode!  
So I've been testing the updated fork, the message is unfortunately not arriving to the app BUT the good news is that XCode is not throwing the previous error anymore, the internal implementation seems to be correct it must be a delegation issue to the app itself...  what is interesting though is that I'm not seeing the merged changes in your repo!! :)

**Comment by CowboyCode** [comment]
True, the commit must have failed, I will recommit it tomorrow

**Comment by 1N50MN14** [comment]
@CowboyCode That would be great thank you!

**Comment by chrisjpalmer** [comment]
Hi guys. I have forked ostownsville's repo as well just like @CowboyCode. I found a bug completely unrelated to this issue to do with copy google-services.json. 
I made a PR on ostownsville's repo BECAUSE it seems to be the new central repo for this plugin.

BUT the reason why I am commenting is because if @CowboyCode is going to put that commit up...

`True, the commit must have failed, I will recommit it tomorrow`

...will you merge with ostownsville?

**Comment by CowboyCode** [comment]
@chrisjpalmer 
Sorry for the confusion, I am ostownsville (work) and also blumanski and CowboyCode, too many computers and locations and ssh keys...

I have already added this pull request, regarding the copy google-services.json to the repo.
https://github.com/fechanique/cordova-plugin-fcm/pull/370

And https://github.com/fechanique/cordova-plugin-fcm/pull/364 (untested)

You can double check it at the moment under
https://github.com/CowboyCode/cordova-plugin-fcm

I will do some testing tomorrow, and then I will merge it into ostownsville and create probably a pull request here.

Cheers

**Comment by 1N50MN14** [comment]
@CowboyCode Just tested it, unfortunately the same error message still appears :(

`Messaging][I-FCM002019] FIRMessaging received data-message, but FIRMessagingDelegate's-messaging:didReceiveMessage: not implemented`

**Comment by chrisjpalmer** [comment]
@CowboyCode Great, thanks for the clarification! Also thank you for contributing to this plugin. We seriously need it!

I can see that you have merged #370 into ostownsville/cordova-plugin-fcm. I actually cloned ostownsville and found a small problem with the way google-services.json is copied. It was only a one liner. I could have it wrong but, the .json started copying properly for me after making the change.

NOW... BACK TO THE ORIGINAL THREAD
I would really love to see the FIRMessaging error go away for iOS! I wish I could help you guys but my iPhone is in repair so I have nothing to test on. Best of luck and thank you!

**Comment by CowboyCode** [comment]
@1N50MN14 @chrisjpalmer 
I merged it now into ostownsville
https://github.com/ostownsville/cordova-plugin-fcm

All is up to date but the error for data messages is still there.
There were breaking changes in version 4.0.0
https://firebase.google.com/docs/reference/ios/firebasemessaging/api/reference/Protocols/FIRMessagingDelegate#-messagingdidreceivemessage

I create a new issue for that.

Cheers

**Comment by chrisjpalmer** [comment]
Hi @CowboyCode could you merge my latest PR?

Thanks

**Comment by CowboyCode** [comment]
Done :-)

**Comment by 1N50MN14** [comment]
@CowboyCode @chrisjpalmer Are you experiencing the plugin requesting push permissions on its own right when it first loads or is it just me?

**Comment by 1N50MN14** [comment]
Ok I see on both Android and iOS the plugin is initialized automatically whenever the webview is ready, then the permission request is triggered; so if there's a splash screen delay this shows up first

**Comment by CowboyCode** [comment]
Regarding the splash screen delay, don't use a splash screen plugin, add it in android itself and you have no problems.

https://www.bignerdranch.com/blog/splash-screens-the-right-way/

**Comment by 1N50MN14** [comment]
@CowboyCode Oh I didn't realize that, thanks for the heads up!

By the way, the init() approach works, I think it's a good idea regardless the splash; in my case I didn't want to show the permission alert first thing when users open the app, only after they have signed up & logged in..

**Comment by 1N50MN14** [comment]
@CowboyCode @chrisjpalmer With the latest update I'm seeing push notifications are no longer showing when the app is closed, was just wondering if you're experiencing a similar issue?

**Comment by CowboyCode** [comment]
I can't test it before Thursday, I let you know then

**Comment by chrisjpalmer** [comment]
Hi CoyboyCode. Troublesome iOS.

I will take a look today.

Regards

Chris

**Comment by 1N50MN14** [comment]
@chrisjpalmer I've been banging my head against the wall with this one, along with the early permission popup which shows the moment the plugin loads (even though I wrapped the `ready` call in the plugin);

**Comment by chrisjpalmer** [comment]
Hi Ayman

The iOS plugin will always display the popup (the way its currently coded), on load. This is because the firebase initialisation routine is fired as soon as the app loads. 
We could incorporate what you are talking about however.

Chris

**Comment by 1N50MN14** [comment]
@chrisjpalmer I think it would be a good idea imho to have a choice on when to initialize firebase, one can choose either on startup or wherever it makes sense.. in my case this has been problematic because the popup is blocking and it appears exactly where I don't want any blocking alerts ;) I'm also a bit concerned Apple may take notice on alerts showing up first thing..

The immediate concern now though is push notifications not showing up on iOS.. it's been stopping me the last few days from pushing the app to the app store :(

**Comment by chrisjpalmer** [comment]
Hi there. Very sorry. Our latest P.R. has broken the plugin I am guessing.

Please revert to a pervious commit by using:
cordova plugin add `https://github.com/ostownsville/cordova-plugin-fcm.git#COMMIT_HASH <https://github.com/ostownsville/cordova-plugin-fcm.git#COMMIT_HASH>` where commit hash is the commit you would like to use of our plugin.

If using Ionic, prefix with ‘ionic’.

Chris

**Comment by 1N50MN14** [comment]
@chrisjpalmer Thank you I'll try to revert 521e288  

One question, the plugin's js automatically invokes https://github.com/ostownsville/cordova-plugin-fcm/blob/master/src/ios/FCMPlugin.m#L26 va https://github.com/ostownsville/cordova-plugin-fcm/blob/master/www/FCMPlugin.js#L40 - i wrapped that inside an `init` call instead, but firebase still somehow managed to run, do you think I missed on something?

**Comment by chrisjpalmer** [comment]
Hi there. Revert to a8e14ad7691cade14288d302b18e7a4ccb7b0c8a.

I don't think the ready method is linked to anything in code. It probably needs to be removed. If you check out https://github.com/ostownsville/cordova-plugin-fcm/blob/master/src/ios/AppDelegate%2BFCMPlugin.m you'll see that the initialisation routine `- (BOOL)application:(UIApplication *)application customDidFinishLaunchingWithOptions:(NSDictionary *)launchOptions {` sets up all the firebase stuff and will thus trigger the popup.

**Comment by 1N50MN14** [comment]
(thank you) right so I see a void method `load` here https://github.com/ostownsville/cordova-plugin-fcm/blob/master/src/ios/AppDelegate%2BFCMPlugin.m#L26 but I don't see where it's being called is it some sort of constructor?

**Comment by chrisjpalmer** [comment]
Yes so checkout this forum post on the Objective-C load method.

https://stackoverflow.com/questions/13326435/nsobject-load-and-initialize-what-do-they-do

load() is called by the runtime. load() performs method swizzling which allows us to hook into the app-delegate's applicationDidFinishLaunching method. Then we can init firebase.

**Comment by 1N50MN14** [comment]
Great, so from the code base point of view what could stop us from having custom initialization? of what I see this is the recommended approach for plugins https://cordova.apache.org/docs/en/latest/guide/platforms/android/plugin.html#plugin-initialization-and-lifetime

**Comment by chrisjpalmer** [comment]
Hi there. So I have tested the plugin this morning on iOS 10.3.3.

**Notifications Not Working**
1) I tested commit a8e14ad to see if the old version worked. At first it didn't. This was because i recently changed my firebase app details over and APNS and Firebase had cached my device id for the old firebase app. Upon deletion of the application and restarting the phone, both foreground and background notifications started working. I was able to see the notification dialog pop up and hear the sounds.

2) I tested the most recent commit of the plugin for comparison. It also works => both foreground and background notifications. I am able to hear sounds and see the dialog.

I should stress that in between tests I...
1. Deleted the app
2. Rebooted the phone
3. Recompiled the app with the plugin version I wanted to test
4. Tested notifications using the Firebase Console notifications test feature
5. Tested silent and non-silent notifications using my server backend.

Conclusion: Notifications appear to be working (at least from my perspective).

**Custom Initialisation**
I personally think this is a really good idea. Thanks for the documentation from the cordova website. If you would like to have a crack at implementing it then go ahead. Otherwise I might have to postpone this until some future time as I am busy with work.

Hope this helps

Chris

**Comment by 1N50MN14** [comment]
@chrisjpalmer I finally figured out the issue! It was a bug in iOS 11!!! Upgraded to 11.1 and push notifications started to arrive,  damn it Apple!!! Thank you so much for the help!!

Going to look into the init thing now

**Comment by raghuyt** [comment]
@1N50MN14 Have you got 11.1 or 11.0.1 in my device i have 11.0.2 but this plugin not working

**Comment by 1N50MN14** [comment]
@raghuyt hmm I just upgraded to 11.0.2, no incoming notifications anymore... 

On 11.0.1 I was receiving notifications, but when the app was closed / in background mode, notification data was not being forwarded to the app..

Only @chrisjpalmer or @CowboyCode can rescue us on this one

**Comment by CowboyCode** [comment]
I work at the moment on ios adk upgrade 4.3.0. may that will help.
But have testing todo, will take a few days.

**Comment by 1N50MN14** [comment]
@CowboyCode Happy to help with testing if you wish

**Comment by chrisjpalmer** [comment]
@1N50MN14, in regards to 

> but when the app was closed / in background mode, notification data was not being forwarded to the app..

Do you mean that your app was not being "woken up" using the silent background notification feature which apple advertise in their documentation. This is meant to be so your application can do a little processing before you open it, making it more performant. 

If you are referring to this, and you are not talking about notification banners (the iOS notification center)... then I have had no success with this either. I don't think it works with APNS anymore.
We tested this in one of our PRs and you can see the results here: https://github.com/ostownsville/cordova-plugin-fcm/pull/4

However our plugin (on iOS 10.3.3 this can be witnessed) does receive the background notification on boot (Point 2.1 in the above link).

Now, it occurred to me the other day that this may be due to the following entitlement not being ticked in XCode, I am actually willing to test this to see if we made this mistake:
![image](https://user-images.githubusercontent.com/29791291/31525039-3632dd76-b009-11e7-8eef-3547236bc74d.png)

Kind Regards
Chris

**Comment by 1N50MN14** [comment]
@chrisjpalmer Yeah I was referring to the notification banner, i.e. the notification showing up on the os notification screen.. the "boot" issue I worked around it by sending the same notification twice, both as normal notification as well as data notification and let the app handle it. 

When it comes to the notification banner, I couldn't find a workaround either, basically users tap the notification, app opens up, they don't see it in the app so they don't know where to go to open the post (my notifications are like "christ published a new post", goes into an app page where lists the notifications).

Everything else worked just fine until 11.0.2 

PS: In addition to the Remote notification entitlement I also have "Background fetch" checked, is this needed for our plugin?

**Comment by raghuyt** [comment]
@1N50MN14  thank you now its working in all IOS versions ,till now IOS version 11.0.2 its working all devices
what i made changes is this
updated our cordova to new version and platform IOS to new version

**Comment by CowboyCode** [comment]
@1N50MN14 
You could test the new firebase sdk update for ios, I have it on
https://github.com/CowboyCode/cordova-plugin-fcm

That would be great :-)

**Comment by 1N50MN14** [comment]
@CowboyCode On it, I'll test it in a bit (about couple of hours) and report back

**Comment by 1N50MN14** [comment]
@CowboyCode Ok I can confirm not only that it's working now but also tapping on the notification from the notification relays the data to the app as well, two problems solved at once!! woohoo!!!! 👍 

// @chrisjpalmer @raghuyt

**Comment by CowboyCode** [comment]
Great, I do the pull request now :-)

**Comment by 1N50MN14** [comment]
@CowboyCode The only thing which is still problematic is notifications not arriving the app when in background mode; but at least notifications are arriving and users are being altered, now I'm thinking to solve this by saving outgoing notifications in the database and load them from there when the app resumes.. a bit of a hassle but at least it's a solution, I'm just happy this plugin works :)

**Comment by chrisjpalmer** [comment]
Hev everyone nice work. @1N50MN14 I know this might not be much help... 

But we found that even though background notifications don't work, if your notification was a silent notification (ie. nothing in the `notification` object in the payload to firebase), the next time the app is opened, the notification will come through. If however your notification was _loud_, opening the app the next time would do nothing.

**Comment by CowboyCode** [comment]
I have just tested the background notifications on ios 11.0.3, that was working fine.
Silent messages worked fine as well.

In my test, I was sending notifications to the devices using the token, not subscribed channels.
@1N50MN14 @chrisjpalmer

**Comment by 1N50MN14** [comment]
I'm experiencing exactly the same case as @chrisjpalmer my initial workaround was whenever I send a loud notification I send it as data message as well, the problem is that in my specific case I'm using AWS SNS to handle notification sends out, and for topics SNS (unlike with direct SNS endpoints) it expects the message payload to be a regular push notification otherwise it'll throw an error.

That leaves me with two options, either drop topics and send messages one by one, or persist the notifications; database writes versus reads, reads are lighter, writes allow to persist both require some coding, opting for writes at least if a user installs the app on a different device I will be able to recover previous notifications and show them in the notification section of the app.. 

But yeah, you're right about silent data notifications; I'm just "unlucky" with my use case..

// @CowboyCode

**Comment by savanr** [comment]
@1N50MN14 @CowboyCode I am having issue when app is in foreground on Android device. Heads up notification doesn't work.

```
{
  priority: 2,
  notification: {
    body: "Good Morning!",
    icon: "myicon",
    sound: 'default'
  }
}
```

**Comment by Simbaclaws** [comment]
I am experiencing the same issue as @savanr. I'm only recieving notifications on android when it is in either the background or killed. When the app is in the foreground I recieve no notifications at all....

Does anyone know why push notifications isn't working when the app is in the foreground?

**Comment by chrisjpalmer** [comment]
Hi guys. I am working on a fork of this project with Cowboy code located here:
https://github.com/ostownsville/cordova-plugin-fcm

Give this a try. If you have any issues, post it on our repository.

**Comment by Simbaclaws** [comment]
Apparently recieving push notifications in the foreground is not possible through firebase cloud messaging. What is possible though is sending data messages together with your notifications and handling the data to show some sort of modal when a notification is recieved in the foreground. I'm currently working on a cordova plugin to play the native push notification sound when it is recieved in the foreground. This way I can recreate the experience of getting a push notification from within my app through data messages send with the notification.

**Comment by ourarash** [comment]
@Simbaclaws : is this the case for both Android and ios? I had the same issue for [android], but don't have ios implementation yet.

(https://stackoverflow.com/questions/47858447/fcm-notification-in-system-tray-when-app-is-in-foreground).

**Comment by Simbaclaws** [comment]
@ourarash I haven't tested it on iOS yet, but will soon in the future. I do think this is the same though. I've read about it somewhere and can confirm this is the case for android. I'll let you know if this is the case for iOS aswell but I'm pretty certain it's how it is intended to work.

**Comment by luigi37** [comment]
Hi all, I can confirm it's both the case in iOS and Android: foreground notification you need to handle the "data" part to issue any type of result as no notification is sent. I think this is by design in FCM.
Please note that this fork contains the most up to date version:
https://github.com/ostownsville/cordova-plugin-fcm
